### PR TITLE
Hints

### PR DESCRIPTION
**Pull request by differentmatt**

### CODE REVIEW

File: app/views/play/level/HintsView.coffee [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    @$el.toggleClass('hide', @hintsState.get('hidden'))
+    super()
+
+  getProcessedHint: ->
+    language = @session.get('codeLanguage')
+    hint = @state.get('hint')
+    return unless hint
+
+    # process
+    translated = utils.i18n(hint, 'body')
```
Comment (phoenixeliot) [inline app/views/play/level/HintsView.coffee]: Does this error if hint is undefined?

File: app/views/play/level/HintsView.coffee [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  incrementHintViewTime: =>
+    hintIndex = @state.get('hintIndex')
+    hintsViewTime = @state.get('hintsViewTime')
+    hintsViewTime[hintIndex] ?= 0
+    hintsViewTime[hintIndex]++
+    hintsUsed = @state.get('hintsUsed')
+    if hintsViewTime[hintIndex] > @hintUsedThresholdSeconds and not hintsUsed[hintIndex] 
+      window.tracker?.trackEvent 'Hint Used', category: 'Students', levelSlug: @level.get('slug'), hintCount: @hintsState.get('hints')?.length ? 0, hintCurrent: hintIndex, ['Mixpanel']
+      hintsUsed[hintIndex] = true
+      @state.set('hintsUsed', hintsUsed)
```
Comment (phoenixeliot) [inline app/views/play/level/HintsView.coffee]: just FYI there's an awesome shorthand if you use the object notation for setting traits:

`@state.set { hintsUsed }`
(equivalent to `@state.set { hintsUsed: hintsUsed }`

File: app/views/play/level/tome/SpellPaletteView.coffee [hunk 3]
```diff
@@ -12,7 +11,7 @@ N_ROWS = 4
 
 module.exports = class SpellPaletteView extends CocoView
   id: 'spell-palette-view'
-  template: template
+  template: require 'templates/play/level/tome/spell-palette-view'
```
Comment (phoenixeliot) [inline app/views/play/level/tome/SpellPaletteView.coffee]: (I like doing templates this way more than at the top) 👍

File: app/views/play/level/tome/SpellPaletteView.coffee [hunk 4]
```diff
@@ -24,13 +23,8 @@ module.exports = class SpellPaletteView extends CocoView
...
 
-  constructor: (options) ->
-    super options
-    @level = options.level
-    @session = options.session
-    @supermodel = options.supermodel
-    @thang = options.thang
-    @useHero = options.useHero
+  initialize: (options) ->
+    {@level, @session, @supermodel, @thang, @useHero} = options
```
Comment (phoenixeliot) [inline app/views/play/level/tome/SpellPaletteView.coffee]: FYI Another useful shorthand if you never need to access `options` as a whole: 
```
initialize: ({@level, @session, @supermodel, @thang, @useHero}) ->
 ```

File: app/views/play/level/HintsView.coffee (hunk 1 above)
Comment (differentmatt) [inline app/views/play/level/HintsView.coffee]: Yea, that's why I return a couple lines up.

**Pull request closed without merging**